File: lua/HTTP.lua
local HttpService = game:GetService("HttpService")
local Promise = require(script.Parent.Promise)
local Settings = require(script.Parent.Settings)

local function GetFullUrl(urlPath)
	local fullUrl = "https://" .. Settings.settings.titleId .. Settings._internalSettings.productionUrl .. urlPath
	local getParams = Settings._internalSettings.requestGetParams
	local firstParam = true
	for key, value in pairs(getParams) do
		if firstParam then
			fullUrl ..= "?"
			firstParam = false
		else
			fullUrl ..= "&"
		end

		fullUrl ..= key .. "=" .. value
	end

	return fullUrl
end

local MakePlayFabApiCall = Promise.promisify(function(path, requestBody, authKey, authValue)
	local fullUrl = GetFullUrl(path)
	local encodedBody = HttpService:JSONEncode(requestBody)
	local headers = {
		["X-ReportErrorAsSuccess"] = "true",
		["X-PlayFabSDK"] = Settings._internalSettings.sdkVersionString,
		["Content-Type"] = "application/json",
	}

	if authKey and authValue ~= "" and authValue then
		headers[authKey] = authValue
	end

	local success, response = pcall(HttpService.RequestAsync, HttpService, {
		Url = fullUrl,
		Method = "POST",
		Headers = headers,
		Body = encodedBody,
	})

	if success then
		if response.Success then
			local responseBody = HttpService:JSONDecode(response.Body)
			if responseBody and responseBody.code == 200 and responseBody.data then
				return responseBody.data
			else
				error(response.StatusMessage)
			end
		else
			error(response.StatusMessage)
		end
	else
		error(tostring(response))
	end
end)

return {
	GetFullUrl = GetFullUrl,
	MakePlayFabApiCall = MakePlayFabApiCall,
}

File: src/HTTP.ts
/// <reference path="PlayFab.d.ts" />

/**
 * Handles sending HTTP requests to PlayFab
 */

import { HttpService } from '@rbxts/services'
import settings from './Settings'

// Types
type TAuthKeys = 'X-EntityToken' | 'X-Authorization' | 'X-SecretKey'

/**
 * Constructs full URL string (from an endpoint path) for use in HTTP requests.
 * 
 * @param urlPath Endpoint path
 */
export const GetFullUrl = (urlPath: string) => {
    let fullUrl = `https://${settings.settings.titleId}${settings._internalSettings.productionUrl}${urlPath}`

    const getParams = settings._internalSettings.requestGetParams
    let firstParam = true
    Object.keys(getParams).forEach(key => {
        const value = getParams[key]
        if (firstParam) {
            fullUrl += '?'
            firstParam = false
        } else fullUrl += '&'

        fullUrl += key + '=' + value
    })

    return fullUrl
}

/**
 * Makes HTTP request to PlayFab.
 * 
 * @param path Endpoint path
 * @param requestBody Body of the request
 * @param authKey The type of authentication
 * @param authValue Authentication value for the specified type
 */
const MakePlayFabApiCall = async<T>(path: string, requestBody: object, authKey?: TAuthKeys, authValue?: string) => {
    const fullUrl = GetFullUrl(path)
    const encodedBody = HttpService.JSONEncode(requestBody)

    // Build request headers
    const headers = new Map<string, string>()
    headers.set('X-ReportErrorAsSuccess', 'true')
    headers.set('X-PlayFabSDK', settings._internalSettings.sdkVersionString)
    headers.set('Content-Type', 'application/json')
    if (authKey && authValue) headers.set(authKey, authValue)

    let [success, response] = pcall(() => {
        return HttpService.RequestAsync({
            Url: fullUrl,
            Method: 'POST', // PlayFab only uses POST, for some reason, don't ask me
            Headers: headers as unknown as HttpHeaders, // types are annoying sometimes
            Body: encodedBody
        })
    })

    if (success) {
        response = response  as RequestAsyncResponse

        if (response.Success) {
            const responseBody = HttpService.JSONDecode(response.Body) as PlayFabModule.IPlayFabResultCommon
            
            if (responseBody && responseBody.code === 200 && responseBody.data) {
                return responseBody.data as T
            } else {
                throw response.StatusMessage
            }
        } else {
            throw response.StatusMessage
        }
    } else {
        throw tostring(response)
    }
}

export default MakePlayFabApiCall